### PR TITLE
Create reference lookup tables

### PR DESCRIPTION
**Pull request by mattwiller**
Creating lookup tables for references between resources, to be populated via a data migration that performs a full reindex of all resource types

### CODE REVIEW

File: packages/server/src/fhir/lookups/reference.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    const values = getSearchReferences(resource);
+    const existing = await getExistingValues(client, this.getTableName(resource.resourceType), resource.id as string);
+    if (compareArrays(values, existing)) {
+      // Nothing changed
+      return;
+    }
+    if (existing.length > 0) {
+      await this.deleteValuesForResource(client, resource);
+    }
+    await this.insertValuesForResource(client, resource.resourceType, values, true);
```
Comment (codyebberson) [inline packages/server/src/fhir/lookups/reference.ts:40]: If we call `deleteValuesForResource`, do we still need to pass `true` for `ignoreConflict`?  Did you find a case where the `ignoreConflict` option was necessary?

File: packages/server/src/fhir/sql.ts [hunk 2]
```diff
@@ -572,15 +572,21 @@ export class ArraySubquery implements Expression {
 
 export class InsertQuery extends BaseQuery {
   private readonly values: Record<string, any>[];
-  private merge?: boolean;
+  private onConflict: 'merge' | 'ignore' | 'error';
```
Comment (codyebberson) [inline packages/server/src/fhir/sql.ts:575]: Can you help me understand the case where `ignore` was necessary?

File: packages/server/src/fhir/lookups/reference.ts (hunk 1 above)
Comment (mattwiller) [reply packages/server/src/fhir/lookups/reference.ts:40]: I think the context got lost when I split this out from the original PR: https://github.com/medplum/medplum/pull/3251#discussion_r1387011984

Comment (codyebberson) [reply packages/server/src/fhir/lookups/reference.ts:40]: Hm - Ok, I think we should fix that either inside `insertValuesForResource` or elsewhere when building the list.  We should use a `Set<string>` to ensure there are no duplicate values.

**Pull request merged**